From 2d5e1fcef2f2c775179b29098834126660861941 Mon Sep 17 00:00:00 2001
From: Archana Patni <archana.patni@intel.com>
Date: Tue, 10 Nov 2015 16:43:15 +0530
Subject: [PATCH 0825/2104] HACK: serial: 8250_dw: skip lpss pci UART D3 when
 used with no_console_suspend

For Platforms with lpss pci UARTs, the parent device should
be prevented from going into D3 for the no_console_suspend
flag to work as expected.

Signed-off-by: Archana Patni <archana.patni@intel.com>
Signed-off-by: Subramony Sesha <subramony.sesha@intel.com>

Interim. This solution is not ideal. We should not be
concerned of the parent in this driver, and we really should
not do any PCI operations here.

Ideally the issue can be fixed in the parent driver
(drivers/mfd/intel-lpss-pci.c).

Signed-off-by: Heikki Krogerus <heikki.krogerus@linux.intel.com>
---
 drivers/tty/serial/8250/8250_dw.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/tty/serial/8250/8250_dw.c b/drivers/tty/serial/8250/8250_dw.c
index 27c5b2b..99fdbb2 100644
--- a/drivers/tty/serial/8250/8250_dw.c
+++ b/drivers/tty/serial/8250/8250_dw.c
@@ -24,9 +24,11 @@
 #include <linux/platform_device.h>
 #include <linux/slab.h>
 #include <linux/acpi.h>
+#include <linux/pci.h>
 #include <linux/clk.h>
 #include <linux/reset.h>
 #include <linux/pm_runtime.h>
+#include <linux/console.h>
 
 #include <asm/byteorder.h>
 
@@ -607,6 +609,20 @@ static int dw8250_suspend(struct device *dev)
 {
 	struct dw8250_data *data = dev_get_drvdata(dev);
 
+	/*
+	 * FIXME: For Platforms with LPSS PCI UARTs, the parent device should
+	 * be prevented from going into D3 for the no_console_suspend flag to
+	 * work as expected.
+	 */
+	if (platform_get_resource_byname(to_platform_device(dev),
+					 IORESOURCE_MEM, "lpss_dev")) {
+		struct uart_8250_port *up = serial8250_get_port(data->line);
+		struct pci_dev *pdev = to_pci_dev(dev->parent);
+
+		if (pdev && !console_suspend_enabled && uart_console(&up->port))
+			pdev->dev_flags |= PCI_DEV_FLAGS_NO_D3;
+	}
+
 	serial8250_suspend_port(data->line);
 
 	return 0;
-- 
2.7.4

